addons/controllers/csi/csiconfig_controller.go

```diff
@@ -0,0 +1,196 @@
+type CSIConfigReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=csi.tanzu.vmware.com,resources=csiconfigs,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=csi.tanzu.vmware.com,resources=csiconfigs/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=csi.tanzu.vmware.com,resources=csiconfigs/finalizers,verbs=update
+//+kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters,verbs=get;list;watch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the CSIConfig object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (r *CSIConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx).WithValues("CSIConfig", req.NamespacedName)
+
+	csiConfig := &csiv1alpha1.CSIConfig{}
+	if err := r.Get(ctx, req.NamespacedName, csiConfig); err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Info("CSIConfig resource not found", "namespacedName", req.NamespacedName)
+			return ctrl.Result{}, nil
+		}
+
+		logger.Error(err, "Unable to fetch CSIConfig resource", "namespacedName", req.NamespacedName)
+		return ctrl.Result{}, err
+	}
+
+	if cluster, err := r.getOwnerCluster(ctx, csiConfig); cluster == nil {
+		return ctrl.Result{}, err // no need to requeue if cluster is not found
+	}
+
+	return r.reconcileCSIConfig(ctx, csiConfig)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *CSIConfigReconciler) SetupWithManager(_ context.Context, mgr ctrl.Manager,
+	options controller.Options) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&csiv1alpha1.CSIConfig{}).
+		WithOptions(options).
+		Complete(r)
+}
+
+func (r *CSIConfigReconciler) reconcileCSIConfig(ctx context.Context,
+	csiCfg *csiv1alpha1.CSIConfig) (result ctrl.Result, retErr error) {
+
+	logger := log.FromContext(ctx)
+
+	patchHelper, err := clusterapipatchutil.NewHelper(csiCfg, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		if retErr != nil {
+			// don't modify CSIConfig if there is an error
+			return
+		}
+
+		if err := patchHelper.Patch(ctx, csiCfg); err != nil {
+			logger.Error(err, "Failed to patch CSIConfig")
+			return
+		}
+	}()
+
+	if !csiCfg.GetDeletionTimestamp().IsZero() {
+		return ctrl.Result{}, nil // deleted
+	}
+
+	if result, err = r.reconcileCSIConfigNormal(ctx, csiCfg); err != nil {
+		logger.Error(err, "Error reconciling CSIConfig")
+		return result, err
```

Comment:
On error you should return empty object ```ctrl.Result{}```

Reply:
reconcileCSIConfigNormal makes sure that result is the appropriate value. It's a delegate function, so we trust it will do the right thing & thus return its results.